README.md

```diff
@@ -1,40 +1,93 @@
 
-## Development tools
+## Why DBSP?
 
-Ideally this code should run fine in Linux, MacOs, and Windows.
-The code is written in Rust.  Here are some tools we found useful for development:
+Computing over streaming data is hard.  Streaming computations operate over
+changing inputs and must continuously update their outputs as new inputs arrive.
+They must do so in real time, processing new data and producing outputs with a
+bounded delay.
 
-See <https://www.twelve21.io/getting-started-with-rust-on-windows-and-visual-studio-code> about
-using Rust in VSCode.
+Existing streaming engines have limited expressive power, which is why their use
+in the modern data stack is restricted to data preparation and visualization.
+Complex business intelligence tasks are left to traditional database systems
+where they run as periodic (e.g., daily) batch jobs.  This is inadequate in the
+world that increasingly relies on real-time alerts, recommendations, control, etc.
 
-* Rust `curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh`
-* Tools for Rust
-  * tarpaulin for code coverage (Linux only): `cargo install cargo-tarpaulin`
-* Vscode: <https://code.visualstudio.com/Download>
-* Vscode extensions for Rust:
-  * crates <https://marketplace.visualstudio.com/items?itemName=serayuzgur.crates>
-  * Rust <https://marketplace.visualstudio.com/items?itemName=rust-lang.rust>
-  * Even Better TOML <https://marketplace.visualstudio.com/items?itemName=tamasfe.even-better-toml>
-  * Power Header <https://marketplace.visualstudio.com/items?itemName=epivision.vscode-file-header>
-    (Enables you to insert a copyright header in VSCode by pressing CTRL+ALT+H)
-  * Rust Test Explorer <https://marketplace.visualstudio.com/items?itemName=swellaby.vscode-rust-test-adapter>
-  * Rust debugger CodeLLDB <https://marketplace.visualstudio.com/items?itemName=vadimcn.vscode-lldb>
-  * Coverage Gutters <https://marketplace.visualstudio.com/items?itemName=ryanluker.vscode-coverage-gutters>
+DBSP intends to change the situation by offering a simple promise: to **execute
+any query workload that can run in batch mode in streaming mode**.
 
-## Check test coverage
+## Overview
 
-* Run on the command line `cargo tarpaulin --out Lcov` to generate coverage information
-* Use the Coverage Gutters VSCode extension to watch the lcov file (CTRL+SHIFT+P, Coverage Gutters: Watch)
+DBSP queries are composed of the following classes of operators (in the order of
+increasing complexity):
 
-## Set-up git hooks
+1. **Per-record operators** that parse, validate, filter, transform data streams
+one record at a time.
+
+1. **Windowing operators** that group time series data into time windows,
+including various forms of tumbling and sliding windows, windows driven by
+watermarks, etc.
+
+1. The complete set of **relational operators**: select, project, join,
+aggregate, etc.
+
+1. **Recursion**: Recursive queries allow for instance expressing graph
```

Comment:
Explanation goes right into example (graph queries) rather than giving a general explanation like you do for {window, per-record} operators.